### PR TITLE
#7783: Port ttnn.upsample to C++

### PR DESCRIPTION
**Pull request by xanderchin**

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/others.hpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace ttnn {
+namespace operations {
+namespace others {
+
+
+inline ttnn::Tensor upsample(
+    const ttnn::Tensor& input_tensor,
+    //std::variant<float, std::tuple<float, float>, std::tuple<float, float, float>, std::tuple<float, float, float, float>> scale_factor,
+    std::variant<int, std::vector<int>> scale_factor,
```
Comment (cfjchu) [inline ttnn/cpp/ttnn/operations/others.hpp:17]: why not change this to use std::variant<std::array<float, 2>, std::array<float, 3>, std::array<float, 4>> and std::visit the helper implementation

File: ttnn/cpp/ttnn/operations/others.hpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            scale_w = sf.at(2);
+            int scale_c = sf.at(3);
+            TT_FATAL(scale_n == 1);
+            TT_FATAL(scale_c == 1);
+        }
+        else {
+            TT_THROW("Unsupported scale factor");
+        }
+    }
+    // DEBUG
```
Comment (cfjchu) [inline ttnn/cpp/ttnn/operations/others.hpp:54]: delete

**Review by cfjchu — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/others.hpp (hunk 1 above)
Comment (cfjchu) [reply ttnn/cpp/ttnn/operations/others.hpp:17]: then you avoid the runtime overhead of the vector size checks

Comment (xanderchin) [reply ttnn/cpp/ttnn/operations/others.hpp]: Missed this comment earlier. Makes sense.. I'm gonna update.

**Pull request merged**